File: Darts/src/burlton/dartzee/code/bean/ScrollTablePlayers.java
package burlton.dartzee.code.bean;

import burlton.core.code.obj.HandyArrayList;
import burlton.dartzee.code.db.PlayerEntity;
import burlton.desktopcore.code.bean.ScrollTable;
import burlton.desktopcore.code.util.TableUtil;

import javax.swing.*;
import java.util.ArrayList;

public class ScrollTablePlayers extends ScrollTable
{
	public ScrollTablePlayers()
	{
		super();
		
		initTableModel();
		
		setRowName("player");	
	}
	
	
	public PlayerEntity getSelectedPlayer()
	{
		int row = table.getSelectedRow();
		if (row == -1)
		{
			return null;
		}
		
		return getPlayerEntityForRow(row);
	}
	public HandyArrayList<PlayerEntity> getAllPlayers()
	{
		HandyArrayList<PlayerEntity> ret = new HandyArrayList<>();
		
		for (int i=0; i<table.getRowCount(); i++)
		{
			PlayerEntity player = getPlayerEntityForRow(i);
			ret.add(player);
		}
		
		return ret;
	}
	public HandyArrayList<PlayerEntity> getSelectedPlayers()
	{
		HandyArrayList<PlayerEntity> ret = new HandyArrayList<>();
		
		int[] viewRows = table.getSelectedRows();
		for (int i=0; i<viewRows.length; i++)
		{
			PlayerEntity player = getPlayerEntityForRow(viewRows[i]);
			ret.add(player);
		}
		
		return ret;
	}
	public PlayerEntity getPlayerEntityForRow(int row)
	{
		//Apparently we don't need to do this conversion
		//int internalRow = table.convertRowIndexToModel(row);
		return (PlayerEntity)table.getValueAt(row, 1);
	}
	
	private void initTableModel()
	{
		initTableModel(new ArrayList<>());
	}
	public void initTableModel(java.util.List<PlayerEntity> players)
	{
		TableUtil.DefaultModel model = new TableUtil.DefaultModel();
		model.addColumn("");
		model.addColumn("Player");
		
		setModel(model);
		
		setRowHeight(23);
		setColumnWidths("25");
		
		for (PlayerEntity player : players)
		{
			addPlayer(player);
		}
		
		sortBy(1, false);
	}
	public void addPlayer(PlayerEntity player)
	{
		ImageIcon flag = player.getFlag();
		Object[] row = {flag, player};
		
		addRow(row);
	}
	
	public void addPlayers(ArrayList<PlayerEntity> players)
	{
		for (PlayerEntity player : players)
		{
			addPlayer(player);
		}
	}
}

File: Darts/src/burlton/dartzee/code/bean/ScrollTablePlayers.kt
package burlton.dartzee.code.bean

import burlton.dartzee.code.db.PlayerEntity
import burlton.dartzee.code.core.bean.ScrollTable
import burlton.dartzee.code.core.util.TableUtil

fun ScrollTable.getSelectedPlayer(): PlayerEntity?
{
    val row = table.selectedRow
    return if (row == -1) null else getPlayerEntityForRow(row)
}

fun ScrollTable.getAllPlayers(): MutableList<PlayerEntity>
{
    val ret = mutableListOf<PlayerEntity>()

    for (i in 0 until table.rowCount)
    {
        val player = getPlayerEntityForRow(i)
        ret.add(player)
    }

    return ret
}


fun ScrollTable.getSelectedPlayers(): MutableList<PlayerEntity>
{
    val ret = mutableListOf<PlayerEntity>()

    val viewRows = table.selectedRows
    for (i in viewRows.indices)
    {
        val player = getPlayerEntityForRow(viewRows[i])
        ret.add(player)
    }

    return ret
}

private fun ScrollTable.getPlayerEntityForRow(row: Int): PlayerEntity
{
    return table.getValueAt(row, 1) as PlayerEntity
}

fun ScrollTable.initPlayerTableModel(players: List<PlayerEntity> = listOf())
{
    val model = TableUtil.DefaultModel()
    model.addColumn("")
    model.addColumn("Player")

    this.model = model

    setRowHeight(23)
    setColumnWidths("25")

    addPlayers(players)

    setRowName("player")
    sortBy(1, false)
}

fun ScrollTable.addPlayers(players: List<PlayerEntity>) = players.forEach{ addPlayer(it) }
private fun ScrollTable.addPlayer(player: PlayerEntity)
{
    val flag = player.getFlag()
    val row = arrayOf(flag, player)

    addRow(row)
}